### PR TITLE
feat(vault): add vault doc

### PR DESCRIPTION
**Pull request by ahmet2mir**
Hello,

This is the documentation part of https://github.com/smallstep/certificates/pull/798 PR

Requires:
- https://github.com/smallstep/certificates/pull/798 PR

### CODE REVIEW

**Comment by maraino** [comment]
# Create a PKI in vault

1. Login to vault, if you are not using https you will need to export an environment variable `export VAULT_ADDR='http://127.0.0.1:8200'`:
```console
vault login token=xxxxx
```

1. Create Root CA:
```console
vault secrets enable pki
vault secrets tune -max-lease-ttl=87600h pki
vault write pki/root/generate/internal common_name="Test Root CA" ttl=87600h
vault read -format=json pki/cert/$(vault list -format=json pki/certs | jq -r '.[0]') | jq -r .data.certificate > root_ca.crt
vault write pki/config/urls issuing_certificates="http://localhost:8200/v1/pki/ca" crl_distribution_points="http://localhost:8200/v1/pki/crl"
```

2. Create subordinate:
```console
vault secrets enable -path=pki_int1 pki
vault secrets tune -max-lease-ttl=43800h pki_int1
vault write -format=json pki_int1/intermediate/generate/internal common_name="Vault Intermediate CA 1" ttl=43800h  | jq -r .data.csr > int1.csr
vault write -format=json pki/root/sign-intermediate csr=@int1.csr format=pem_bundle ttl=43800h | jq -r .data.certificate > int1.crt
cat int1.crt root_ca.crt > intermediate_ca.crt
vault write pki_int1/intermediate/set-signed certificate=@intermediate_ca.crt
vault write pki_int1/config/urls issuing_certificates="http://127.0.0.1:8200/v1/pki_int1/ca" crl_distribution_points="http://127.0.0.1:8200/v1/pki_int1/crl"
```

3. Create roles:
```console
vault write pki_int1/roles/ec-role key_type=ec allow_any_name=true enforce_hostnames=false max_ttl=72h
vault write pki_int1/roles/rsa-role key_type=rsa allow_any_name=true enforce_hostnames=false max_ttl=72h
vault write pki_int1/roles/ed25519-role key_type=ed25519 allow_any_name=true enforce_hostnames=false max_ttl=72h
```

4. Create policy (I'm not sure about what you really need here):
```console
echo 'path "pki_int1/sign/ec-role" {
    capabilities = ["create", "update"]
}
path "pki_int1/sign/rsa-role" {
    capabilities = ["create", "update"]
}
path "pki_int1/sign/ed25519-role" {
    capabilities = ["create", "update"]
}
path "pki_int/revoke" {
    capabilities = ["create", "update"]
}' > policy.hcl
vault policy write smallstep policy.hcl
```

5. Configure authentication:
```console
$ vault auth enable -description="SmallStep RA AppRole" approle
Success! Enabled approle auth method at: approle/
$ vault write auth/approle/role/approle \
     bind_secret_id=true \
     local_secret_ids=false \
     secret_id_num_uses=0 \
     secret_id_ttl=0s \
     token_explicit_max_ttl=0s \
     token_max_ttl=0s \
     token_no_default_policy=false \
     token_num_uses=0 \
     token_period=0s \
     token_policies="smallstep" \
     token_ttl=0s \
     token_type=default
Success! Data written to: auth/approle/role/approle
$ vault read -field=role_id auth/approle/role/approle/role-id
xxxxxx
$ vault write -field=secret_id -f auth/approle/role/approle/secret-id
yyyyyy
```

Your `ca.json` will contain something like:
```json
{
  "...": "...",
  "authority": {
    "type": "vaultcas",
    "certificateAuthority": "http://127.0.0.1:8200",
    "certificateAuthorityFingerprint": "foobar...",
    "config": {
      "pki": "pki_int1",
      "pkiRoleRSA": "rsa-role",
      "pkiRoleEC": "ec-role",
      "pkiRoleEd25519": "ed25519-role",
      "appRole": "approle",
      "roleID": "xxxxxx",
      "secretID": {
        "fromString": "yyyyyy"
      },
      "IsWrappingToken": false
    },
    "...": "..."
  }
}
```

**Comment by brotherdust** [comment]
@ahmet2mir please advise: do you have a docs style guide for contributors that I can follow? I want to make sure any changes I suggest comply with the general style and approach.

Thanks!

**Comment by tashian** [comment]
Hi @brotherdust I'm a Developer Advocate at Smallstep and I maintain a lot of the docs here. [The README for this repo](https://github.com/smallstep/docs#readme) has some info on style and syntax, and how to preview the docs locally. And, once this is ready I'll do a final round of proofreading, style checking, etc. Thanks for your contribution!

**Comment by ahmet2mir** [comment]
> @ahmet2mir please advise: do you have a docs style guide for contributors that I can follow? I want to make sure any changes I suggest comply with the general style and approach.
> 
> Thanks!

I try to follow the same doc than Google CAS
Will appreciate if you could follow the doc to check if it works as expected and new eyes is always better.

@maraino what the purpose of your comment ? A resume of commands you used and to replace in the doc ?

**Comment by maraino** [comment]
> what the purpose of your comment ? A resume of commands you used and to replace in the doc ?

@ahmet2mir I haven't looked at your docs, my comments are the steps I used to create my PKI, with vault docs it's not straightforward for a user how to configure the full PKI. It can be used almost as a script to start a new one, I'm also using `jq` and output redirections to write locally the necessary certificates or CSRs, and combining them to upload the bundles. The steps are not straightforward if you just look at Vault docs. 

Basically, it was a reference that we can add to the docs, and it was more for @tashian to create the final version or if he wanted to try.

**Comment by tashian** [comment]
Here's a jq incantation that will add vaultcas stuff to ca.json:

```
VAULT_URI=http://127.0.0.1:8200
VAULT_CA_FINGERPRINT=$(step certificate fingerprint root_ca.crt)
VAULT_ROLE_ID=$(vault read -field=role_id auth/approle/role/approle/role-id)
VAULT_SECRET_ID=$(vault write -field=secret_id -f auth/approle/role/approle/secret-id)

cat <<< $(jq --arg vault_role_id "$VAULT_ROLE_ID" \
            --arg vault_secret_id "$VAULT_SECRET_ID" \
            --arg vault_uri "$VAULT_URI" \
            --arg vault_ca_fingerprint "$VAULT_CA_FINGERPRINT" \
             '(.authority += {
                "type": "vaultcas",
                "certificateAuthority": $vault_uri,
                "certificateAuthorityFingerprint": $vault_ca_fingerprint,
                "config": {
                          "pki": "pki_int1",
                          "pkiRoleRSA": "rsa-role",
                          "pkiRoleEC": "ec-role",
                          "pkiRoleEd25519": "ed25519-role",
                          "appRole": "approle",
                          "roleID": $vault_role_id,
                          "secretID": {
                                "fromString": $vault_secret_id
                          },
                          "IsWrappingToken": false
                }
    })' $(step path)/config/ca.json) > $(step path)/config/ca.json
```

**Comment by tashian** [comment]
Moved to #92.
